art.hpp

```diff
@@ -49,52 +49,54 @@ using node_ptr = basic_node_ptr<node_header>;
 
 struct impl_helpers;
 
-template <typename Key>
-using inode_defs = basic_inode_def<inode<Key>, inode_4<Key>, inode_16<Key>,
-                                   inode_48<Key>, inode_256<Key>>;
+template <typename Key, typename Value>
+using inode_defs = basic_inode_def<inode<Key, Value>, inode_4<Key, Value>,
+                                   inode_16<Key, Value>, inode_48<Key, Value>,
+                                   inode_256<Key, Value>>;
 
-template <typename Key, class INode>
-using db_inode_deleter = basic_db_inode_deleter<INode, unodb::db<Key>>;
+template <typename Key, typename Value, class INode>
+using db_inode_deleter = basic_db_inode_deleter<INode, unodb::db<Key, Value>>;
 
-template <typename Key>
+template <typename Key, typename Value>
 using art_policy =
-    basic_art_policy<Key, unodb::db, unodb::in_fake_critical_section,
+    basic_art_policy<Key, Value, unodb::db, unodb::in_fake_critical_section,
                      unodb::fake_lock, unodb::fake_read_critical_section,
                      node_ptr, inode_defs, db_inode_deleter,
                      basic_db_leaf_deleter>;
 
-template <typename Key>
-using inode_base = basic_inode_impl<art_policy<Key>>;
+template <typename Key, typename Value>
+using inode_base = basic_inode_impl<art_policy<Key, Value>>;
 
 template <typename Key>
 using leaf_type = basic_leaf<Key, node_header>;
 
-template <typename Key>
-class inode : public inode_base<Key> {};
+template <typename Key, typename Value>
+class inode : public inode_base<Key, Value> {};
 
 }  // namespace detail
 
-template <typename Key>
+template <typename Key, typename Value>
 class mutex_db;
 
 /// A non-thread-safe implementation of the Adaptive Radix Tree (ART).
 ///
 /// See olc_art for a highly concurrent thread-safe ART
 /// implementation.
-template <typename Key>
+template <typename Key, typename Value>
 class db final {
-  friend class mutex_db<Key>;
+  friend class mutex_db<Key, Value>;
 
  public:
   using key_type = Key;
+  using value_type = Value;
   using value_view = unodb::value_view;
   using get_result = std::optional<value_view>;
```

Comment:
Should we change the get_result type to std::optional<value_type> so we have simple return by value semantics for integral values?

We should keep value_view anyway as a special type that we understand which is used for variable length values.